Docs: Continued work on extension documentation for DI (#21807)
backstage/backstage: docs/frontend-system/architecture/03-extensions.md

```tsx
const extension = createExtension({
  // ...
  output: {
    element: coreExtensionData.reactElement.optional(),
  },
  factory() {
    return {
      element: <img src="./assets/logo.png" />,
    };
  },
});
```

## Extension Inputs

The Extension Data can be passed up to other extensions through their extension inputs. Similar to the outputs seen before, let's create an example an extension with a extension input:

```tsx
const navigationExtension = createExtension({
  // ...
  inputs: {
    // [1]: Input
    logo: createExtensionInput(
      {
        element: coreExtensionData.reactElement,
      },
      { singleton: true, optional: true },
    ),
  },
  factory({ inputs }) {
    return {
      element: (
        <nav>{inputs.logo.output?.element ?? <span>Backstage</span>}</nav>
      ),
    };
  },
  // ...
});
```

> The input (see [1] above) is an object with the properties `$$type`, `extensionData` and `config`, where `createExtensionInput` helps creating this object by taking the `extensionData` as the first parameter, and the `config` as the second parameter making sure the `$$type` is set correctly. The extension data ensures the input is of the type & format we expect. With `config` the format & type can be defined even more precisely. The extension input can be converted from an array of inputs to just a single input by setting `singleton` to `true`. If the extension input is optional, `optional: true` converts the type to be optional as well.

Reviewer: ```suggestion
The input (see [1] above) is an object that we create using `createExtensionInput`. The first argument is the set of extension data that we accept via this input, and works just like the `output` option. The second argument is optional, and it allows us to put constraints on the extensions that are attached to our input. If the `singleton: true` option is set, only a single extension can attached at a time, and unless the `optional: true` option is set it will also be required that there is exactly on attached extension.
```

The `$$type` is more of an internal thing that I don't think we should mention in docs tbh, unless we specifically describe our system for opaque types somewhere. I updated the rest a bit to to try to highlight the difference between input shape and shape of attached extensions

```tsx
const navigationItemExtension = createExtension({
  // ...
  attachTo: { id: 'app/nav', input: 'items' },
  factory() {
    return {
      element: <Link to="/home">Home</Link>,
    };
  },
});

const navigationExtension = createExtension({
  // ...
  // [2]: Extension `id` will be `app/nav` following the extension naming pattern
  namespace: 'app',
  name: 'nav',
  output: {
    element: coreExtensionData.reactElement,
  },
  inputs: {
    items: createExtensionInput({
      element: coreExtensionData.reactElement,
    }),
  },
  factory({ inputs }) {
    return {
      element: (
        <nav>
          <ul>
            {inputs.items.map(item => {
              return <li>{item.output.element}</li>;
            })}
          </ul>
        </nav>
      ),
    };
  },
  // ...
});
```

In this case the extension input `items` is an array, where each individual item is an extension that attached itself to the extension inputs of this `id`.

> With the `inputs` not only the `output` of an extensions item is passed to the extension, but also the `node`. It is discouraged to consume and/or modify the `node` here. If we are looking at the `factory` function from the example above we could access the `node` like the following:

Reviewer: ```suggestion
With the `inputs` not only the `output` of an extensions item is passed to the extension, but also the `node`. However, it is discouraged to consume the `node` here unless needed. If we are looking at the `factory` function from the example above we could access the `node` like the following:
```

The node should absolutely not be modified 😁

```tsx
  // ...
  factory({ inputs }) {
    return {
      element: (
        <nav>
          <ul>
            {inputs.items.map(({output, node}) => {
              const _node: AppNode = node;
              return <li>{output.element}</li>;
            })}
          </ul>
        </nav>
      ),
    };
  },
```

## Extension Configuration

> With the `app-config.yaml` there is already the option to pass configuration to plugins or the app to e.g. define the `baseURL` of your app. For extensions this concept would be limiting as an extension can be independet of the plugin & initiated several times. Therefor we created a possibility to configure each extension indiviually through config. The config schema is created using the library `zod` (TODO: Link), which enables a stronger type checking on top of TypeScript. If we continue with the example of the `navigationExtension` and now want it to contain a configurable title, we could make it available like the following:

Reviewer: ```suggestion
With the `app-config.yaml` there is already the option to pass configuration to plugins or the app to e.g. define the `baseURL` of your app. For extensions this concept would be limiting as an extension can be independet of the plugin & initiated several times. Therefor we created a possibility to configure each extension indiviually through config. The extension config schema is created using the [`zod`](https://zod.dev/) library, which in addition to TypeScript type checking also provides runtime validation and coercion. If we continue with the example of the `navigationExtension` and now want it to contain a configurable title, we could make it available like the following:
```

```tsx
const navigationExtension = createExtension({
  // ...
  namespace: 'app',
  name: 'nav',
  // [3]: Extension `id` will be `app/nav` following the extension naming pattern
  configSchema: createSchemaFromZod(z =>
    z.object({
      title: z.string().default('Sidebar Title'),
    }),
  ),
  factory({ config }) {
    return {
      element: (
        <nav>
          <span>{config.title}</span>
          <ul>{/* ... */}</ul>
        </nav>
      ),
    };
  },
  // ...
});
```

To now change the text of the title from "Sidebar Title" to "Backstage" we can look at the `id` of the extension & add the following to the `app-config.yaml`:

```yaml
app:
  # ...
  extensions:
    # ...
    - app/nav:
        config:
          title: 'Backstage'
```

> For further information on how to write documentation please follow the architecture section on this (TODO: Link).

Reviewer: ```suggestion
<!-- For further information on how to write documentation please follow the architecture section on this (TODO: Link). -->
```

Now sure what documentation this is talking about though?

If we follow the example from above all items of the navigation have similarites, like they all want to be attached to the same extension with the same input as well as rendering the same navigation item component. Therefor `createExtension` can be abstracted for this use case to `createNavItemExtension` and if we add the extension to the app it will end up in the right place & looks like we expect a navigation item to look.

```tsx
export const HomeNavIcon = createNavItemExtension({
  routeRef: routeRefForTheHomePage,
  title: 'Home',
  icon: HomeIcon,
});
```

### Extension Kind

With the example `HomeNavIcon` will end up on the extension input `items` of the extensions with the id `app/nav`. It raises the question what the `id` of the `HomeNavIcon` itself is. The extension creator for the navigation item has a defined `kind`, which by convention matches the own name. So in this example `createNavItemExtension` sets the kind to `nav-item`.

The `id` of the extension is then build out of `namespace`, `name` & `kind` like the following - where `namespace` & `name` are optional properties that can be passed to the extension creator:

```
id: kind:namespace/name
```

> For more information on naming of extension refer to the naming patterns documentation (TODO: Link)

Reviewer: ```suggestion
For more information on naming of extension refer to the [naming patterns documentation](./08-naming-patterns.md).
```

### Extension Creators in libraries

Extension creators should be exported from library packages (e.g. `*-react`, `*-common`) rather than plugin packages.

<!--

Explain that extension creators should be exported from library packages (`-react`) rather than plugin packages.

> philipph: What is the advantage - not super obvious to me from the comment. We would also need to change this e.g. for the catalog as it currently is exported from `catalog` alpha directly.

Reviewer: It's a bit more nuanced that what the comment implies. Extension creators that are intended for use by other plugins or modules should be exported from the `-react` package. Never `-common` though since we can't have web-only things there.

There are situations where exporting from the plugin package itself might be alright, primarily that's for when a particular extension kind is really only intended as an internal customization. Still, if a plugin has a `-react` package it's usually best to place the creators there, and yep the catalog filter creator should be moved over imo.

We should probably say something to the effect of "if you only intend for an extension kind to be used for internal customizations and not by other open source plugins, then you can place the extension creator in the plugin package itself. But if you already have a `-react` package or intend for the extension creator to be used by other open source plugins, then you should always put extension creators in a `-react` package.
